lib/cfg.c

```diff
@@ -226,6 +221,33 @@ cfg_load_forced_modules(GlobalConfig *self)
     }
 }
 
+static gboolean
+_is_module_autoload_enabled(GlobalConfig *self)
+{
+  return atoi(cfg_args_get(self->globals, "autoload-compiled-modules") ? : "1");
+}
+
+void
+cfg_discover_candidate_modules(GlobalConfig *self)
+{
+  if (self->use_plugin_discovery && _is_module_autoload_enabled(self))
+    {
+      if (_sync_plugin_module_path_with_global_define(self) || !self->plugin_discovery_finished)
+        plugin_discover_candidate_modules(&self->plugin_context);
+      self->plugin_discovery_finished = TRUE;
```

Comment:
Hello,
 we discussed this IRL with @Kokan , and we have a proposal to eliminate the new `plugin_discovery_finished` variable. Handling global path change as a "plugin_context invalidation" event.

```
if (_sync_plugin_module_path_with_global_define(self))
  {
    _free_candidate_plugins(&self->plugin_context);
    self->plugin_context = NULL;
  }

if (self->plugin_context == NULL)
  plugin_discover_candidate_modules(&self->plugin_context);
```

note: This code should not call `_free_candidate_plugins` with a NULL value, but I would add a NULL check, like g_free has.


We also came up with an idea to move this discovery out of the lexer's scope, but we rather merge this PR, than widening the scope any further.

Reply:
This way we would be calling a private function from a higher layer (_free_candidate_plugins), which is not a good idea I think.

Also, plugin_context is not a pointer right now. I am not sure why that extra variable is bad, but if you think it is better, then a plugin_has_discovery_run() function could achieve something similar.

Reply:
I have implemented this in the latest incarnation.